src/native/aws_signing.c

```diff
@@ -0,0 +1,325 @@
+        jni_headers,
+        jni_body_stream);
+
+done:
+
+    if (jni_uri != NULL) {
+        (*env)->DeleteLocalRef(env, jni_uri);
+    }
+
+    if (jni_headers != NULL) {
+        (*env)->DeleteLocalRef(env, jni_headers);
+    }
+
+    return http_request;
+}
+
+static void s_complete_signing_exceptionally(
+    JNIEnv *env,
+    struct s_aws_sign_request_callback_data *callback_data,
+    int error_code) {
+
+    jint jni_error_code = error_code;
+    struct aws_byte_cursor error_cursor = aws_byte_cursor_from_c_str(aws_error_name(error_code));
+    jstring jni_error_string = aws_jni_string_from_cursor(env, &error_cursor);
+    AWS_FATAL_ASSERT(jni_error_string);
+
+    jobject crt_exception = (*env)->NewObject(
+        env,
+        crt_runtime_exception_properties.crt_runtime_exception_class,
+        crt_runtime_exception_properties.constructor_method_id,
+        jni_error_code,
+        jni_error_string);
+    AWS_FATAL_ASSERT(crt_exception);
+
+    (*env)->CallBooleanMethod(
+        env, callback_data->java_future, completable_future_properties.complete_exceptionally_method_id, crt_exception);
+
+    (*env)->DeleteLocalRef(env, jni_error_string);
+}
+
+static void s_aws_signing_complete(struct aws_signing_result *result, int error_code, void *userdata) {
+
+    struct s_aws_sign_request_callback_data *callback_data = userdata;
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    if (result == NULL || error_code != AWS_ERROR_SUCCESS) {
+        s_complete_signing_exceptionally(
+            env, callback_data, (error_code != AWS_ERROR_SUCCESS) ? error_code : AWS_ERROR_UNKNOWN);
+        goto done;
+    }
+
+    if (aws_apply_signing_result_to_http_request(callback_data->native_request, aws_jni_get_allocator(), result)) {
+        s_complete_signing_exceptionally(env, callback_data, aws_last_error());
+        goto done;
+    }
+
+    jobject java_signed_request =
+        s_create_signed_java_http_request(env, callback_data->native_request, callback_data->java_original_request);
+    if (java_signed_request == NULL) {
+        s_complete_signing_exceptionally(env, callback_data, aws_last_error());
+        goto done;
+    }
+
+    (*env)->CallBooleanMethod(
+        env, callback_data->java_future, completable_future_properties.complete_method_id, java_signed_request);
+
+done:
+
+    s_cleanup_callback_data(callback_data);
+}
+
+static bool s_should_sign_param(const struct aws_byte_cursor *name, void *user_data) {
+    (void)name;
+
+    struct s_aws_sign_request_callback_data *callback_data = user_data;
+
+    JNIEnv *env = aws_jni_get_thread_env(callback_data->jvm);
+
+    jstring parameter_name = aws_jni_string_from_cursor(env, name);
+
+    bool result = (*env)->CallBooleanMethod(
+        env, callback_data->java_sign_param_predicate, predicate_properties.test_method_id, (jobject)parameter_name);
+
+    (*env)->DeleteLocalRef(env, parameter_name);
+
+    return result;
+}
+
+static int s_build_signing_config(
```

Comment:
verify ref semantics here.

Reply:
The result is passed into the signing function in the direction of Java -> Native.  When combined with the fact that the signing config is copied internally at the start of signing, the automatic local ref destruction that happens on return up the call stack exactly matches the reference life time we need.